flocker/node/script.py

```diff
@@ -281,143 +320,312 @@ def get_service(self, reactor, options):
-    """
-    def main(self, reactor, options, volume_service):
-        agent_config = options[u'agent-config']
-        configuration = yaml.safe_load(agent_config.getContent())
+    get_external_ip = field(initial=_get_external_ip, mandatory=True)
+
+    control_service_host = field(type=bytes, mandatory=True)
+    control_service_port = field(type=int, mandatory=True)
+
+    # Cannot use type=NodeCredential because one of the tests really wants to
+    # set this to None.
+    node_credential = field(mandatory=True)
+    # Cannot use type=Certificate; pyrsistent rejects classic classes.
+    ca_certificate = field(mandatory=True)
+
+    backend_name = field(type=unicode, mandatory=True)
+    api_args = field(type=PMap, factory=pmap, mandatory=True)
+
+    @classmethod
+    def from_configuration(cls, configuration):
+        """
+        Load configuration from a data structure loaded from the configuration
+        file and only minimally processed.
+
+        :param dict configuration: Agent configuration as returned by
+            ``get_configuration``.
 
-        validate_configuration(configuration=configuration)
+        :return: A new instance of ``cls`` with values loaded from the
+            configuration.
+        """
+        host = configuration['control-service']['hostname']
+        port = configuration['control-service']['port']
+
+        node_credential = configuration['node-credential']
+        ca_certificate = configuration['ca-certificate']
+
+        api_args = configuration['dataset']
+        backend_name = api_args.pop('backend')
 
-        deployer_factory = dataset_deployer_from_configuration(
-            dataset_configuration=configuration['dataset'],
-            volume_service=volume_service
+        return cls(
+            control_service_host=host,
+            control_service_port=port,
+
+            node_credential=node_credential,
+            ca_certificate=ca_certificate,
+
+            backend_name=backend_name.decode("ascii"),
+            api_args=api_args,
         )
 
-        service_factory = AgentServiceFactory(
-            deployer_factory=deployer_factory
-        ).get_service
+    def get_backend(self):
+        """
+        Find the backend in ``self.backends`` that matches the one named by
+        ``self.backend_name``.
 
-        service = service_factory(reactor, options)
+        :raise ValueError: If ``backend_name`` doesn't match any known backend.
+        :return: The matching ``BackendDescription``.
+        """
+        for backend in self.backends:
+            if backend.name == self.backend_name:
+                return backend
+        raise ValueError(
+            "Backend named {} not available".format(self.backend_name),
+        )
 
-        if configuration['dataset']['backend'] == 'zfs':
-            # XXX This should not be a special case,
-            # see https://clusterhq.atlassian.net/browse/FLOC-1924.
-            volume_service.setServiceParent(service)
```

Comment:
The service parent primarily ensures that `startService` and `stopService` are called at appropriate times. Since there's no `stopService` it seemed appropriate to just call `startService` at the start instead. This allows the `startService` call to be done before the convergence loop service is created, meaning it fits in with the rest of the ZFS backend initialization - no more special case for ZFS after the convergence loop is started.